src/engine/TransitivePath.cpp

```diff
@@ -215,361 +218,168 @@ size_t TransitivePath::getCostEstimate() {
+                    subRes->idTable(), _lhs, _rhs, leftRes->idTable());
 
-  const IdTableView<SUB_WIDTH> sub = dynSub.asStaticView<SUB_WIDTH>();
-  const IdTableView<LEFT_WIDTH> left = dynLeft.asStaticView<LEFT_WIDTH>();
-  IdTableStatic<RES_WIDTH> res = std::move(*dynRes).toStatic<RES_WIDTH>();
+  } else if (_rhs.treeAndCol.has_value()) {
+    LOG(DEBUG) << "TransitivePath right result computation..." << std::endl;
+    shared_ptr<const ResultTable> rightRes =
+        _rhs.treeAndCol.value().first->getResult();
+    LOG(DEBUG) << "TransitivePath right result computation done." << std::endl;
+    size_t rightWidth = rightRes->idTable().numColumns();
 
-  // Used to map entries in the left column to entries they have connection with
-  Map edges;
+    LOG(DEBUG) << "TransitivePath result computation..." << std::endl;
+    CALL_FIXED_SIZE((std::array{_resultWidth, subWidth, rightWidth}),
+                    &TransitivePath::computeTransitivePathBound, this, &idTable,
+                    subRes->idTable(), _rhs, _lhs, rightRes->idTable());
 
-  auto checkTimeoutAfterNCalls = checkTimeoutAfterNCallsFactory();
-  auto checkTimeoutHashSet = [&checkTimeoutAfterNCalls]() {
-    checkTimeoutAfterNCalls(NUM_OPERATIONS_HASHSET_LOOKUP);
-  };
-  // initialize the map from the subresult
-  for (size_t i = 0; i < sub.size(); i++) {
-    checkTimeoutHashSet();
-    Id l = sub(i, leftSubCol);
-    Id r = sub(i, rightSubCol);
-    MapIt it = edges.find(l);
-    if (it == edges.end()) {
-      std::shared_ptr<ad_utility::HashSet<Id>> s =
-          std::make_shared<ad_utility::HashSet<Id>>();
-      s->insert(r);
-      edges[l] = s;
-    } else {
-      // If r is not in the vector insert it
-      it->second->insert(r);
-    }
+  } else {
+    LOG(DEBUG) << "TransitivePath result computation..." << std::endl;
+    CALL_FIXED_SIZE((std::array{_resultWidth, subWidth}),
+                    &TransitivePath::computeTransitivePath, this, &idTable,
+                    subRes->idTable(), _lhs, _rhs);
   }
 
-  // For every node do a dfs on the graph
-
-  // Stores nodes we already have a path to. This avoids cycles.
-  ad_utility::HashSet<Id> marks;
-
-  // The stack used to store the dfs' progress
-  std::vector<ad_utility::HashSet<Id>::const_iterator> positions;
-  // Used to store all edges leading away from a node for every level.
-  // Reduces access to the hashmap, and is safe as the map will not
-  // be modified after this point.
-  std::vector<std::shared_ptr<const ad_utility::HashSet<Id>>> edgeCache;
+  LOG(DEBUG) << "TransitivePath result computation done." << std::endl;
+  // NOTE: The only place, where the input to a transitive path operation is not
+  // an index scan (which has an empty local vocabulary by default) is the
+  // `LocalVocabTest`. But it doesn't harm to propagate the local vocab here
+  // either.
+  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
```

Comment:
This is probably wrong, as the `left` or `right` side can also have nonempty local vocabs, so you have to call the `getSharedNonEmptyOf (have forgotten the name) method in the `ResultTable` class here  (not your fault, but your chance to fix it).

Reply:
I've looked at the function `ResultTable::getSharedLocalVocabFromNonEmptyOf`, but I don't understand how it is supposed to be called. The docstring says it assumes that one of the result tables is empty, so should I call it like `subRes->getSharedLocalVocabFromNonEmptyOf(leftRes, ResultTable(...))`?